### ISSUE TITLE
Nested if not working

### ISSUE THREAD

**Issue by Feliix42** [ohua-dev/ohua-core#29]
The following example for a nested `if` is not compiling:
```
ns some_ns;

use sf conditionals::iftest::{modify_string_positive, modify_string_negative, get_ctrl_input, get_another_ctrl_input, get_input};

fn main() -> String {
    let ctrl = get_ctrl_input();
    let another_ctrl = get_another_ctrl_input();
    let inp = get_input();

    if (ctrl) {
        if (another_ctrl) {
            modify_string_positive(inp)
        } else {
            modify_string_negative(inp)
        }
    } else {
        inp
    }
}
```

due to an error while lowering to DFLang:
```
[Error] Argument must be local binding or literal, was Lambda (Binding "__1")
(Let (Binding "a") (Apply (Lit (FunRefLit (FunRef (QualifiedBinding (NSRef [Binding
"conditionals",Binding "iftest"]) (Binding "modify_string_positive")) Nothing)))
(Var (Binding "inp_0_0"))) (Var (Binding "a")))
```

[[ALang-core dump](https://hastebin.com/lociwequjo)]

**Comment by sertel** [comment]
The reason for this is that the lowering do not traverse the lambdas that they lift into control. I'm afraid that `smap` suffers from the same problem.
The fix is easy though.

**Comment by sertel** [comment]
The reason why we lift literals is actually to catch the following class of calls:
```
f 1
```
These would not be depending on a `ctrl` otherwise.
But we need a more efficient solution for that because right now it will lift all the literals and we get tons of `nth` nodes.

The removal of the `nth` nodes is actually not working here. Because many `nth` nodes get created just because of literals for the `nth` nodes of the innermost `if`.

**Comment by sertel** [comment]
It would be really great if we could make them `unitFn`s. Then we could just give the `unit` signal once to the `ctrl`.

**Comment by sertel** [comment]
As a matter of fact there is a much simpler solution:
Currently, we do not pay attention during lambda lifting whether this is really an independent function of just a function that takes a literal *and* another arg:
```
f 1
f 1 someVar
```
As soon as we have fixed that we have at the very least a solution for the problem that I see now.

**Comment by Feliix42** [comment]
The same problem actually arises for an `smap` nested into an `if`:

```
ns some_ns;

use sf conditionals::iftest::{get_ctrl_input, get_vec_input, times_2};

fn main() -> Vec<i32> {
    let cond = get_ctrl_input();

    if (cond) {
        let y = get_vec_input();
        for num in y {
            times_2(num)
        }
    } else {
        get_vec_input()
    }
}
```

Error Message:
```
[Error] Argument must be local binding or literal, was Lambda (Binding "num_0") (Let (Binding "a")
(Apply (Lit (FunRefLit (FunRef (QualifiedBinding (NSRef [Binding "conditionals",Binding "iftest"])
(Binding "times_2")) Nothing))) (Var (Binding "num_0"))) (Var (Binding "a")))
```

`alang-resolved`:
```
let cond = conditionals.iftest/get_ctrl_input () in
ohua.lang/if cond
  (λ _ ->
     let y = conditionals.iftest/get_vec_input () in
     ohua.lang/smap (λ num -> conditionals.iftest/times_2 num) y)
  (λ _ -> conditionals.iftest/get_vec_input ())
```

**Comment by JustusAdam** [comment]
@Feliix42 could you try it with the commit that I made?

**Comment by Feliix42** [comment]
Testing with commit adcfc21e71ad20aedead38307be93efbaafc5dfd, the most basic conditionals-testcase starts failing 😞 
```
ns some_ns;

use sf conditionals::iftest::{modify_string_positive, modify_string_negative, get_cond_result, get_input};

fn main() -> String {
    let cond = get_cond_result(); // yields true
    let inp = get_input();

    if (cond) {
        modify_string_positive(inp)
    } else {
        modify_string_negative(inp)
    }
}
```
ALang-resolved:
```
let cond = conditionals.iftest/get_cond_result () in
let inp = conditionals.iftest/get_input () in
ohua.lang/if cond (λ _ -> conditionals.iftest/modify_string_positive inp)
  (λ _ -> conditionals.iftest/modify_string_negative inp)
```

---

```
[Error] Found if with unexpected, non-unit-lambda branch(es)
          true:
           Lambda (Binding "__0") (Let (Binding "a") (Apply (Lit (FunRefLit (FunRef
             (QualifiedBinding (NSRef [Binding "conditionals",Binding "iftest"]) (Binding
             "modify_string_positive")) Nothing))) (Var (Binding "inp_0"))) (Var (Binding "a")))
          false:
          Lambda (Binding "__1") (Let (Binding "b") (Apply (Lit (FunRefLit (FunRef
             (QualifiedBinding (NSRef [Binding "conditionals",Binding "iftest"]) (Binding
             "modify_string_negative")) Nothing))) (Var (Binding "inp_0"))) (Var (Binding "b")))
```

**Comment by JustusAdam** [comment]
I found the problem there it was a condition I implemented incorrectly.

**Comment by Feliix42** [comment]
Okay, basic functionality is restored, but [this one (comment above)](https://github.com/ohua-dev/ohua-core/issues/29#issuecomment-459667032) persists, the file from the comment does not compile with the same error message.

**Comment by JustusAdam** [comment]
When I run `cargo test if` i only see messages of the `not yet implemented` kind.

**Comment by Feliix42** [comment]
Either pull the latest changes or just manually try to build the files with `ohuac`. Relevant files are:
```
testcases/src/conditionals/algorithms/smap_in_if_no_passthrough.ohuac
testcases/src/conditionals/algorithms/smap_in_if_passthrough.ohuac
```

**Comment by JustusAdam** [comment]
Update: Issues in the compiler are fixed as far as I know, as soon as #33 is also implemented in the backend we should be ready to close this.
